### PR TITLE
Fix preview hotkey issue

### LINKED ISSUE #1927: I opened Flow Launcher and it give me this error
[2023-02-22.txt](https://github.com/Flow-Launcher/Flow.Launcher/files/10807980/2023-02-22.txt)

Flow Launcher version: 1.13.0
OS Version: Microsoft Windows NT 10.0.22623.0
IntPtr Length: 8
x64: True

Python Path: C:\Users\Ulvdi\AppData\Roaming\FlowLauncher\Environments\Python\PythonEmbeddable-v3.8.9\pythonw.exe
Node Path: 
Date: 02/22/2023 21:14:38
Exception:
System.NotSupportedException: 'Shift+D1' key and modifier combination is not supported for KeyGesture.
   at System.Windows.Input.KeyGesture..ctor(Key key, ModifierKeys modifiers, String displayString, Boolean validateGesture)
   at System.Windows.Input.KeyGestureConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object source)
   at Flow.Launcher.Converters.StringToKeyBindingConverter.Convert(Object value, Type targetType, Object parameter, CultureInfo culture) in C:\projects\flow-launcher\Flow.Launcher\Converters\StringToKeyBindingConverter.cs:line 15
   at System.Windows.Data.BindingExpression.TransferValue(Object newValue, Boolean isASubPropertyChange)
   at System.Windows.Data.BindingExpression.Activate(Object item)
   at System.Windows.Data.BindingExpression.AttachToContext(AttachAttempt attempt)
   at System.Windows.Data.BindingExpression.MS.Internal.Data.IDataBindEngineClient.AttachToContext(Boolean lastChance)
   at MS.Internal.Data.DataBindEngine.Task.Run(Boolean lastChance)
   at MS.Internal.Data.DataBindEngine.Run(Object arg)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)

### PR DESCRIPTION
**Pull request by VictoriousRaptor**
## Changes

- Check if preview hotkey is a valid KeyGesture when setting and loading
- Ban more printable characters
- Automate fix for #1755.
- Fix #1927.

## Tests

- Invalid preview hotkey in settings.json won't crash.
- Invalid preview hotkey or toggle hotkey can't be set.

### CODE REVIEW

File: Flow.Launcher/ViewModel/MainViewModel.cs [hunk 1]
```diff
@@ -583,7 +584,24 @@ public double MainWindowWidth
...
+        public string PreviewHotkey 
+        { 
+            get
+            {
+                // TODO try to patch issue #1755
+                // Should be removed after some time
+                try
+                {
+                    var converter = new KeyGestureConverter();
+                    var key = (KeyGesture)converter.ConvertFromString(Settings.PreviewHotkey);
```
Comment (taooceros) [inline Flow.Launcher/ViewModel/MainViewModel.cs:596]: probably this is better? [ConvertFromInvariantString(String)](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.typeconverter.convertfrominvariantstring?view=windowsdesktop-7.0#system-componentmodel-typeconverter-convertfrominvariantstring(system-string))

Correct me if i am wrong

Comment (jjw24) [reply Flow.Launcher/ViewModel/MainViewModel.cs:596]: Probably CultureInfo matters here, we are not dealing with datetime, decimal or floats.

File: Flow.Launcher.Infrastructure/Hotkey/HotkeyModel.cs [hunk 2]
```diff
@@ -137,8 +137,13 @@ public override string ToString()
...
         }
-
-        public bool Validate()
+        
+        /// <summary>
+        /// Validate hotkey
+        /// </summary>
+        /// <param name="validateKeyGestrue">Try to validate hotkey as a KeyGesture.</param>
+        /// <returns></returns>
+        public bool Validate(bool validateKeyGestrue = false)
```
Comment (jjw24) [inline Flow.Launcher.Infrastructure/Hotkey/HotkeyModel.cs:146]: Do we need this option? Shouldn't we always validate?

If needed can we set validate to true by default?

File: Flow.Launcher/ViewModel/MainViewModel.cs [hunk 3]
```diff
@@ -583,7 +584,24 @@ public double MainWindowWidth
 
         public string OpenResultCommandModifiers => Settings.OpenResultModifiers;
 
-        public string PreviewHotkey => Settings.PreviewHotkey;
+        public string PreviewHotkey 
+        { 
+            get
+            {
+                // TODO try to patch issue #1755
+                // Should be removed after some time
```
Comment (jjw24) [inline Flow.Launcher/ViewModel/MainViewModel.cs:592]:
```suggestion
                // Added in v1.14.0, remove after v1.16.0. 
```

File: Flow.Launcher/HotkeyControl.xaml.cs [hunk 4]
```diff
@@ -19,6 +19,11 @@ public partial class HotkeyControl : UserControl
 
         public event EventHandler HotkeyChanged;
 
+        /// <summary>
+        /// Designed for Preview Hotkey and KeyGesture.
+        /// </summary>
+        public bool ValidateKeyGesture { get; set; } = false;
```
Comment (jjw24) [inline Flow.Launcher/HotkeyControl.xaml.cs:25]: Same here, should it be true by default?

**Review by jjw24 — APPROVED** [review]

**Pull request merged**